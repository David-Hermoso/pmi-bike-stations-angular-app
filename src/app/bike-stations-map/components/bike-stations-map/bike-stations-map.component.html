<google-map
  class="map"
  width="100%"
  height="100%"
  [center]="centerCoords"
  [zoom]="13"
>
  <map-marker
    *ngFor="let bikeStation of bikeStations | async"
    [position]="{
    lat: bikeStation.lat,
    lng: bikeStation.lng
    }"
    [title]="bikeStation.name"
    (mapClick)="onMarkerClick(bikeStation)"
  ></map-marker>
</google-map>
