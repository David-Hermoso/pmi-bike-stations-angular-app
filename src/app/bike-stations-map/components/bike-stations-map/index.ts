export * from './bike-stations-map.component';
