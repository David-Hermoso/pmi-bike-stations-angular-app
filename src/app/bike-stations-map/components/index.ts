export * from './bike-stations-map';
