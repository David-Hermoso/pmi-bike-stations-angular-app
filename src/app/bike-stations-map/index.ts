export * from './components';
export * from './bike-stations-map.module';
