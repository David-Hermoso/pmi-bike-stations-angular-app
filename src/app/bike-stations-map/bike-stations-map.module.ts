import {NgModule} from '@angular/core';
import {CommonModule} from '@angular/common';
import {BikeStationsMapComponent} from './components';
import {GoogleMapsModule} from '@angular/google-maps';
import {BikeStationInfoDrawerComponent} from './components/station-info-drawer/bike-station-info-drawer.component';
import {MatSidenavModule} from '@angular/material/sidenav';
import {MatListModule} from '@angular/material/list';
import {MatIconModule} from '@angular/material/icon';
import {MatButtonModule} from '@angular/material/button';
import {RouterModule} from '@angular/router';
import {DrawerActionBtnComponent} from './components/drawer-action-btn/drawer-action-btn.component';
import {MatTooltipModule} from '@angular/material/tooltip';
import {TranslateModule} from '@ngx-translate/core';

@NgModule({
  declarations: [
    BikeStationsMapComponent,
    BikeStationInfoDrawerComponent,
    DrawerActionBtnComponent
  ],
  exports: [
    BikeStationInfoDrawerComponent
  ],
    imports: [
        CommonModule,
        GoogleMapsModule,
        MatSidenavModule,
        MatListModule,
        MatIconModule,
        MatButtonModule,
        RouterModule,
        MatTooltipModule,
        TranslateModule
    ]
})
export class BikeStationsMapModule {
}
