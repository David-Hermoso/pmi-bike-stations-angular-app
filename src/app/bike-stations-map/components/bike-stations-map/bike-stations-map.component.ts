import {Component} from '@angular/core';
import {select, Store} from '@ngrx/store';
import {AppState} from '../../../models/app.state';
import {selectBikeStations} from '../../../state/reducers';
import {Observable} from 'rxjs';
import {BikeStation} from '../../../models/bike-station';
import {showBikeStationInfo} from '../../../state/actions/bike-stations-map.actions';

@Component({
  selector: 'app-bike-stations-map',
  templateUrl: './bike-stations-map.component.html',
  styleUrls: ['./bike-stations-map.component.scss']
})
export class BikeStationsMapComponent {
  public centerCoords = {
    lat: 39.575670,
    lng: 2.654020
  };
  public bikeStations: Observable<BikeStation[]>;

  constructor(private store: Store<AppState>) {
    this.bikeStations = this.store.pipe(select(selectBikeStations));
  }

  public onMarkerClick(bikeStation: BikeStation): void {
    this.store.dispatch(showBikeStationInfo({
      payload: bikeStation
    }));
  }
}
