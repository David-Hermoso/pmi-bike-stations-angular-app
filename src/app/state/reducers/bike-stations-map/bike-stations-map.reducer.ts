import {createReducer, createSelector, on} from '@ngrx/store';
import * as BikeStationsMapActions from '../../actions/bike-stations-map.actions';
import {AppState} from '../../../models/app.state';
import {BikeStation} from '../../../models/bike-station';

export interface BikeStationsMapState {
  selectedBikeStation: BikeStation | undefined;
}

export const initialBikeStationsMapState: BikeStationsMapState = {
  selectedBikeStation: undefined
};

export const selectBikeStationsMapState = (appState: AppState) => appState.bikeStationsMap;

export const selectSelectedBikeStation = createSelector(
  selectBikeStationsMapState,
  ({selectedBikeStation}: BikeStationsMapState) => selectedBikeStation
);

export const bikeStationsMapReducer = createReducer(
  initialBikeStationsMapState,
  on(BikeStationsMapActions.showBikeStationInfo,
    (currentState: BikeStationsMapState, {payload}) => ({
      ...currentState,
      selectedBikeStation: payload
    })
  ),
  on(BikeStationsMapActions.closeBikeStationInfo, (currentState: BikeStationsMapState) => ({
    ...currentState,
    selectedBikeStation: undefined
  }))
);
