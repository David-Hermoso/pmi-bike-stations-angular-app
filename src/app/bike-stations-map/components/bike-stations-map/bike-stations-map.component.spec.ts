import {of} from 'rxjs';
import {BikeStationsMapComponent} from './bike-stations-map.component';

describe('BikeStationsMapComponent', () => {

  let component: BikeStationsMapComponent;

  it('should dispatch showBikeStationInfo', () => {
    const mockStore = {
      dispatch: jest.fn(),
      pipe: jest.fn(() => of([
        {
          img: '003.jpeg',
          type: 'anchor',
          id: '01',
          name: 'PLAÇA DELS PATINS',
          lng: 2.659399509,
          lat: 39.56590061
        }
      ]))
    };
    component = new BikeStationsMapComponent(mockStore as any);
    component.onMarkerClick({
      img: null,
      id: '27',
      name: 'PL. PARIS',
      lng: 2.649158,
      lat: 39.584186,
      fullAddress: 'Plaça de París, 1, 07010 Palma, Illes Balears',
      street: 'Plaça de París',
      streetNumber: 1,
      cp: '07010',
      town: 'Palma',
      region: 'Illes Balears'
    });

    expect(mockStore.dispatch).toHaveBeenCalled();
  });
  it('should set selectedBikeStation', done => {
    const bikeStation = {
      img: null,
      id: '27',
      name: 'PL. PARIS',
      lng: 2.649158,
      lat: 39.584186,
      fullAddress: 'Plaça de París, 1, 07010 Palma, Illes Balears',
      street: 'Plaça de París',
      streetNumber: 1,
      cp: '07010',
      town: 'Palma',
      region: 'Illes Balears'
    };
    const mockStore = {
      dispatch: jest.fn(),
      pipe: jest.fn(() => of(bikeStation)
      )
    };
    component = new BikeStationsMapComponent(mockStore as any);

    mockStore.pipe().subscribe((selectedBikeStation) => {
      expect(selectedBikeStation).toStrictEqual(bikeStation);
      done();
    });
  });

  it('onCloseBikeStationInfo: should dispatch closeBikeStationInfo', () => {
    const bikeStation = {
      img: null,
      id: '27',
      name: 'PL. PARIS',
      lng: 2.649158,
      lat: 39.584186,
      fullAddress: 'Plaça de París, 1, 07010 Palma, Illes Balears',
      street: 'Plaça de París',
      streetNumber: 1,
      cp: '07010',
      town: 'Palma',
      region: 'Illes Balears'
    };
    const mockStore = {
      dispatch: jest.fn(),
      pipe: jest.fn(() => of(bikeStation)
      )
    };
    component = new BikeStationsMapComponent(mockStore as any);
    component.onCloseBikeStationInfo();

    expect(mockStore.dispatch).toHaveBeenCalled();
  });
});
