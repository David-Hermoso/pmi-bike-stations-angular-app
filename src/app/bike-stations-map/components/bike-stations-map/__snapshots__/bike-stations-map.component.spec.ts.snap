// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BikeStationsMapComponent centerCoords should be set 1`] = `
Object {
  "lat": 39.57567,
  "lng": 2.65402,
}
`;
